ruby/extractor/src/diagnostics.rs

```diff
@@ -0,0 +1,281 @@
+            Some(Severity::Note) => tracing::info!("{}", full_error_message),
+            None => tracing::debug!("{}", full_error_message),
+        }
+        if self.inner.is_none() {
+            let mut open_failed = false;
+            self.inner = self.path.as_ref().and_then(|path| {
+                match std::fs::OpenOptions::new()
+                    .create(true)
+                    .append(true)
+                    .write(true)
+                    .open(&path)
+                {
+                    Err(e) => {
+                        tracing::error!(
+                            "Could not open log file '{}': {}",
+                            &path.to_string_lossy(),
+                            e
+                        );
+                        open_failed = true;
+                        None
+                    }
+                    Ok(file) => Some(std::io::BufWriter::new(file)),
+                }
+            });
+            if open_failed {
+                self.path = None
+            }
+        }
+        if let Some(mut writer) = self.inner.as_mut() {
+            serde_json::to_writer(&mut writer, mesg)
+                .unwrap_or_else(|e| tracing::debug!("Failed to write log entry: {}", e));
+            &mut writer
+                .write_all(b"\n")
+                .unwrap_or_else(|e| tracing::debug!("Failed to write log entry: {}", e));
+        }
+    }
+}
+
+pub struct DiagnosticLoggers {
+    extractor: String,
+    root: Option<PathBuf>,
+}
+
+impl DiagnosticLoggers {
+    pub fn new(extractor: &str) -> Self {
+        let env_var = format!(
+            "CODEQL_EXTRACTOR_{}_DIAGNOSTIC_DIR",
+            extractor.to_ascii_uppercase()
+        );
+
+        let root = match std::env::var(&env_var) {
+            Err(e) => {
+                tracing::error!("{}: {}", e, &env_var);
+                None
+            }
+            Ok(dir) => {
+                if let Err(e) = std::fs::create_dir_all(&dir) {
+                    tracing::error!("Failed to create log directory {}: {}", &dir, e);
+                    None
+                } else {
+                    Some(PathBuf::from(dir))
+                }
+            }
+        };
+        DiagnosticLoggers {
+            extractor: extractor.to_owned(),
+            root,
+        }
+    }
+
+    pub fn logger(&self) -> LogWriter {
+        thread_local! {
+            static THREAD_NUM: usize = {
+                static COUNT: std::sync::atomic::AtomicUsize = std::sync::atomic::AtomicUsize::new(0);
+                COUNT.fetch_add(1, std::sync::atomic::Ordering::SeqCst)
+            };
+        }
+        THREAD_NUM.with(|n| LogWriter {
+            extractor: self.extractor.to_owned(),
+            inner: None,
+            path: self
+                .root
+                .as_ref()
+                .map(|root| root.to_owned().join(format!("extractor_{}.jsonl", n))),
+        })
+    }
```

Comment:
Yes, it would, except for the main thread, for which we can map `None -> "main"`.  It would be a bit brittle though if we ever add more non-rayon threads or add multiple thread pools.

Reply:
You're right, that it is not very likely. Having two threads opening (or closing) the same file is likely to cause trouble.